### DISCUSSION TITLE
MacOS: If sound I/O changed after launch, changes not reflected settings

### DISCUSSION THREAD

**Discussion by Skeezixes** [jamulussoftware/jamulus#1065] [Bug Investigation]
Start Jamulus client and open Settings. On my Mac laptop the default options are In: built-in mic Out: built-in speakers. Then, I go to System Preferences and choose options other than these, close the System Preferences window and go back to Jamulus and open Settings. The new settings are not reflected in the dropdown list of options. They remain the same as before. Quit Jamulus client, leaving the settings in System Preferences what you have changed them to. 

Open Jamulus client and Settings. The new options show up in the dropdown list of options. 

If you don't understand what's happening, that Jamulus doesn't update changes to input/output but only fetches the status at launch, it really causes all sorts of confusion and difficulty. Please fix.

**Comment by chrisrimple** [comment]
The problem description is confusing.  It seems to be a duplicate of #711, but may be describing:

- Mac user opens Sys Prefs > Sound and sets Input = Built-In Mic and Output = Built-In Speakers
- User runs Jamulus and sets Device = in:Built-In Mic/out:Built-In Speakers
- User opens Sys Prefs > Sound and sets Input = Apple Airpods and Output = Apple Airpods
- Jamulus continues using Built-In Mic and Built-In Speakers

The request seems to be that changes to Mac System Preferences should result in an automatic change in Jamulus settings.  I disagree, except in this example, which is slightly (but importantly) different from above:

- Mac user opens Sys Prefs > Sound and sets Input = Built-In Mic and Output = Built-In Speakers
- User runs Jamulus and sets Device = **in:System/out:System**
- User opens Sys Prefs > Sound and sets Input = Apple Airpods and Output = Apple Airpods
- Jamulus **switches to use Apple Airpods**

But for that to work, Jamulus may need to be regularly rescanning for audio devices and/or responding to a system message that the System device has changed.

@Skeezixes Have I correctly described your request?

**Comment by Skeezixes** [comment]
Yes. Jamulus wouldn’t have to poll System Settings tho, would it?

**Comment by chrisrimple** [comment]
Don't know, depends on whether MacOS is sending "change" message to which Jamulus can respond when there's a change to System sound settings.  If not, then Jamulus would need to occasionally "scan" for device changes, maybe every 60 seconds when the Setup window has focus (or similar).

**Comment by kwindrem** [comment]
If MacOS doesn't send change notifications, I feel it would be sufficient to update the sound devices when entering the Settings menu.
You could also provide a "Refresh List" as the first item in the list.

**Comment by Skeezixes** [comment]
This sounds like a good solution to me. 

Leigh

**Comment by corrados** [comment]
I have created a Mac beta version of Jamulus where the problem should hopefully be fixed. Can you please test this version is provide me with your feedback: https://github.com/corrados/jamulus/issues/727#issuecomment-732397463

**Comment by gilgongo** [comment]
Is this the issue that was fixed in the 3.6.2? If so I guess we can close?

**Comment by corrados** [comment]
No, this issue is still open. There is a workaround for this issue by selecting any driver from the list. After that the list is refreshed. But it would be better to get a system notification from the MacOS which teills Jamulus that a sound card device change has happened and that we then can refresh the list.

**Comment by ann0see** [comment]
I think this issue is much broader than just on macOS. https://github.com/jamulussoftware/jamulus/commit/d8a5c6fdc7ec9bceaf6fb8569dce1a2962c64597 introduced a bug on Windows. We will probably need re think this and introduce something like a "reload device list" feature on Windows and macOS.

**Comment by gilgongo** [comment]
@ann0see Does this need further discussion or can we turn this ticket into an action as described https://github.com/jamulussoftware/jamulus/issues/181#issuecomment-744690295 (ie system notification)

**Comment by ann0see** [comment]
This issue is still important. No idea yet how/what to do. Somebody more knowledgeable should have a look at this? I'd leave it open

**Comment by gilgongo** [comment]
OK but if it's not an actionable thing that somebody can pick up and reasonably work on, then it should really be in discussions. We need to start assigning issues to projects and milestones at some point (once we've got a road map of some kind).

**Comment by ann0see** [comment]
Yeah ok. So just convert it.

**Comment by chrisrimple** [comment]
I agree that this can move to a conversation.  This problem isn't resolved, but it's better with 3.6.2's "error message on startup if prior audio device is no longer available".  I think there is an open question about whether this issue **should** be resolved - specifically:

1. In Jamulus, user has Device = in:System/out:System
2. User navigates to System Preferences > Sound and selects a different Input or Output (effectively changing System values)
3. Should Jamulus react to that change?

By (1) "listening" for a system message from MacOS that a change has been made, or maybe (2) polling regularly for changes, the Jamulus client **could** know that a change has happened.  But should it react?  I can good reasons yes and no.  However, this feels like a low-frequency use case for which there is an easy workaround (in Jamulus client, change Device to what's desired), so it's a very low-priority request.

**Reply by gilgongo** [reply]
@chrisrimple 
> However, this feels like a low-frequency use case for which there is an easy workaround

OK, so unless anyone else wants to argue for the contrary, we can leave it here and maybe close off eventually otherwise.

**Comment by litong01** [comment]
not sure if this has been resolved, I have a question about similar issue though, i use iPhone client to connect to my local server, connection is fine, then I connect my airpod to my iPhone, this setup is to allow me to sing with my friends over internet, but iphone does not pick up the voice from my airpod, if I disconnect my airpod, jamulus will pick up sound from the iPhone microphone, I wonder how can I make the airpod work as voice input to jamulus client? any hint is very much appreciated.

**Reply by ann0see** [reply]
It will probably introduce a lot of delay. Also the AirPod microphone is as far as I remember quite bad.

**Reply by ann0see** [reply]
I'd buy a 10€ Lightning/USB headphone jack adapter and use wired headphones if you can't use an interface
